qubes/tests/integ/vm_qrexec_gui.py

```diff
@@ -65,6 +65,262 @@ def setUp(self):
             self.testvm2.provides_network = True
         self.app.save()
 
+class TC_00_AudioMixin(TC_00_AppVMMixin):
+    def wait_for_pulseaudio_startup(self, vm):
+        self.loop.run_until_complete(
+            self.wait_for_session(self.testvm1))
+        try:
+            self.loop.run_until_complete(vm.run_for_stdio(
+                "timeout 30s sh -c 'while ! pactl info; do sleep 1; done'"
+            ))
+        except subprocess.CalledProcessError as e:
+            self.fail('Timeout waiting for pulseaudio start in {}: {}{}'.format(
+                vm.name, e.stdout, e.stderr))
+        # then wait for the stream to appear in dom0
+        local_user = grp.getgrnam('qubes').gr_mem[0]
+        p = self.loop.run_until_complete(asyncio.create_subprocess_shell(
+            "sudo -E -u {} timeout 60s sh -c '"
+            "while ! pactl list sink-inputs | grep -q :{}; do sleep 1; done'"
+            .format(local_user, vm.name)))
+        self.loop.run_until_complete(p.wait())
+        # and some more...
+        self.loop.run_until_complete(asyncio.sleep(1))
+
+    def prepare_audio_vm(self, backend):
+        if 'whonix-gw' in self.template:
+            self.skipTest('whonix-gw have no audio')
+        self.loop.run_until_complete(self.testvm1.start())
+        pulseaudio_units = 'pulseaudio.socket pulseaudio.service'
+        pipewire_units = 'pipewire.socket wireplumber.service pipewire.service'
+        is_fedora = 'fedora' in self.template
+        comment = '' if is_fedora else '#'
+        if backend == 'pulseaudio':
+            stop_units = pipewire_units
+            start_units = pulseaudio_units
+            manage_cmd = 'rm -f'
+            packages = 'pulseaudio'
+        elif backend == 'pipewire':
+            if not is_fedora:
+                self.skipTest('PipeWire tests only supported on Fedora')
+            start_units = pipewire_units
+            stop_units = pulseaudio_units
+            manage_cmd = 'touch'
+            packages = 'wireplumber pipewire-pulseaudio'
+        else:
+            self.fail('bad audio backend')
+        try:
+            self.loop.run_until_complete("which parecord")
+        except subprocess.CalledProcessError:
+            self.skipTest('pulseaudio-utils not installed in VM')
+        try:
+            self.testvm1.run_for_stdio(f"""sudo {manage_cmd} /run/qubes-service/pipewire &&
```

Comment:
Will that leave the legacy agent untested?

Reply:
No, legacy agent will get tested by running this test (pulseaudio variant) on a template with either just pulseaudio installed, or both but without pipewire-pulseaudio (since it defaults to pulseaudio then, unless "pipewire" qvm-service is enabled).